File: internal/mcp/tools.go
// Package mcp serves the platform's tool surface.
//
// # The one property this package exists to hold
//
// **Everything tools/list shows, tools/call accepts. Everything tools/call
// accepts, tools/list shows.**
//
// The dangerous half is a tool that appears and then refuses. An AI reads a
// tool list as a menu of things it may do; a tool that is visible and denied
// teaches it that denials are noise to retry through, and an AI that has
// learned to treat denials as noise is one that will keep pushing on the
// denials that matter. Listing something you cannot call is worse than not
// listing it.
//
// The way that guarantee is kept is not care, it is arithmetic: **both paths
// ask the same predicate the same question.** ListTools and CallTool both call
// Guard.ToolReason with the same install and the same tool name. There is no
// second filter, no visibility flag, no "hidden" that means "denied". Two
// predicates that agree today are two predicates, and the failure mode of two
// predicates is exactly the one above.
//
// Manifest-level `hidden` is not an exception. It removes a tool from the
// surface before either path sees it, so it is invisible AND uncallable
// together ... which is the same guarantee, not a violation of it.
package mcp

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"sort"

	"github.com/google/uuid"

	"github.com/bees-roadhouse/hive-sandbox/internal/identity"
	"github.com/bees-roadhouse/hive-sandbox/internal/manifest"
	"github.com/bees-roadhouse/hive-sandbox/internal/trust"
)

// Tool is one entry in tools/list, as an MCP client sees it.
type Tool struct {
	// Name is qualified by app, so tools/call can route by prefix and two
	// installs cannot both offer `list`.
	Name        string         `json:"name"`
	Description string         `json:"description,omitempty"`
	InputSchema map[string]any `json:"inputSchema,omitempty"`
}

// Install is one installed app in some actor's scope, with its derived surface.
type Install struct {
	ID      uuid.UUID
	App     string
	Surface manifest.Surface
}

// Installs yields the ACTIVE installs an actor could be offered tools from.
//
// "Could be offered" rather than "may call": this is the candidate set, and
// every candidate still goes through the predicate. A source that pre-filtered
// by permission would be a second enforcement point, and the first place the
// two would disagree is the bug this package exists to prevent.
type Installs interface {
	ActiveInstalls(ctx context.Context, cred identity.Credential) ([]Install, error)
}

// Guard decides whether an actor may call one tool of one install.
//
// One method, used by both paths, deliberately. An interface with a cheap
// "can see" and an expensive "can call" would be an invitation to use the cheap
// one for listing, and the two would drift.
//
// It also carries the audit obligation: an access allowed only by an override
// writes an audit row inside this call (D18.2), which is why listing audits
// too. Listing a tool you can only see through an override IS an override
// access, and making it free would put the obligation back on one call site.
type Guard interface {
	ToolReason(ctx context.Context, cred identity.Credential, installID uuid.UUID, tool string) (Allowed, error)
}

// Allowed is the guard's verdict, reduced to what this package needs.
type Allowed bool

// Dispatcher runs a tool once the guard has said yes.
//
// It never decides permission. Handing it a credential and letting it check
// would be invariant 11's mistake: a check that takes as an argument the fact
// it is deciding about.
type Dispatcher interface {
	// CallGuest invokes an exported guest function.
	CallGuest(ctx context.Context, in GuestCall) (Result, error)
	// CallCRUD runs a generated operation host-side, with no guest involved.
	CallCRUD(ctx context.Context, in CRUDCall) (Result, error)
}

// GuestCall is one invocation of a guest function.
type GuestCall struct {
	Install  Install
	Cred     identity.Credential
	Function string
	Input    json.RawMessage
}

// CRUDCall is one generated operation.
type CRUDCall struct {
	Install    Install
	Cred       identity.Credential
	Collection string
	Op         manifest.Op
	Input      json.RawMessage
}

// Result is what a tool produced.
type Result struct {
	Output json.RawMessage
	// Trust rides out with the result, because an MCP client is frequently an
	// AI and this is the value that decides whether the content may reach
	// instruction position (invariant 9).
	Trust trust.Level
	// TaintedBy names what first made the invocation untrusted, so "why is this
	// untrusted" is answerable without reading a log.
	TaintedBy string
}

// Errors a caller can classify without matching on prose.
var (
	// ErrUnknownTool is a name no install in the actor's scope offers. It is
	// deliberately the same answer as "you may not call that": distinguishing
	// them tells an unauthorized caller which tools exist.
	ErrUnknownTool = errors.New("mcp: no such tool")

	// ErrMalformedName is a tool name that is not app-qualified.
	ErrMalformedName = errors.New("mcp: tool name is not qualified")
)

// Server answers tools/list and tools/call for one platform.
type Server struct {
	installs   Installs
	guard      Guard
	dispatcher Dispatcher
}

// New builds a server. All three collaborators are required: a nil guard would
// be a server that lists everything, which is the failure this package is about.
func New(installs Installs, guard Guard, dispatcher Dispatcher) (*Server, error) {
	if installs == nil || guard == nil || dispatcher == nil {
		return nil, errors.New("mcp: installs, guard and dispatcher are all required")
	}
	return &Server{installs: installs, guard: guard, dispatcher: dispatcher}, nil
}

// ListTools is the union over installs in the connecting actor's scope,
// filtered by the same predicate tools/call uses (D2.1).
//
// Sorted, so a client diffing two listings sees real changes rather than map
// order.
func (s *Server) ListTools(ctx context.Context, cred identity.Credential) ([]Tool, error) {
	if err := cred.Validate(); err != nil {
		return nil, err
	}
	installs, err := s.installs.ActiveInstalls(ctx, cred)
	if err != nil {
		return nil, err
	}

	var out []Tool
	for _, inst := range installs {
		for _, tool := range inst.Surface.Tools {
			ok, err := s.guard.ToolReason(ctx, cred, inst.ID, tool.Name)
			if err != nil {
				return nil, fmt.Errorf("mcp: authorize %s.%s: %w", inst.App, tool.Name, err)
			}
			if !ok {
				continue
			}
			out = append(out, Tool{
				Name:        manifest.QualifiedToolName(inst.App, tool.Name),
				Description: tool.Description,
				InputSchema: tool.InputSchema,
			})
		}
	}
	sort.Slice(out, func(i, j int) bool { return out[i].Name < out[j].Name })
	return out, nil
}

// CallTool routes a qualified name into a guest or a generated operation.
//
// The lookup and the permission check are the same two steps ListTools takes,
// in the same order, against the same predicate. That is what makes the two
// agree, and it is why this does not consult a cached listing: a listing is a
// snapshot, and a grant revoked between listing and calling has to bite.
func (s *Server) CallTool(
	ctx context.Context, cred identity.Credential, name string, input json.RawMessage,
) (Result, error) {
	if err := cred.Validate(); err != nil {
		return Result{}, err
	}
	app, tool, ok := manifest.SplitToolName(name)
	if !ok {
		return Result{}, fmt.Errorf("%w: %q", ErrMalformedName, name)
	}

	installs, err := s.installs.ActiveInstalls(ctx, cred)
	if err != nil {
		return Result{}, err
	}

	for _, inst := range installs {
		if inst.App != app {
			continue
		}
		for _, t := range inst.Surface.Tools {
			if t.Name != tool {
				continue
			}
			allowed, err := s.guard.ToolReason(ctx, cred, inst.ID, t.Name)
			if err != nil {
				return Result{}, fmt.Errorf("mcp: authorize %s: %w", name, err)
			}
			if !allowed {
				// Same error as "no such tool", on purpose. Telling an
				// unauthorized caller that a tool exists is an existence
				// oracle, and the tool list they can see already tells them
				// everything they are allowed to know.
				return Result{}, fmt.Errorf("%w: %q", ErrUnknownTool, name)
			}

			switch t.Impl {
			case manifest.ImplGeneratedCRUD:
				return s.dispatcher.CallCRUD(ctx, CRUDCall{
					Install: inst, Cred: cred,
					Collection: t.Collection, Op: t.Op, Input: input,
				})
			default:
				return s.dispatcher.CallGuest(ctx, GuestCall{
					Install: inst, Cred: cred, Function: t.Function, Input: input,
				})
			}
		}
	}
	return Result{}, fmt.Errorf("%w: %q", ErrUnknownTool, name)
}

File: crates/hive-mcp/tests/tools.rs
//! The tool surface, ported from internal/mcp/tools_test.go.

use std::collections::HashMap;
use std::sync::Arc;

use async_trait::async_trait;
use hive_identity::{Credential, PrincipalKind};
use hive_manifest::*;
use hive_mcp::*;
use hive_trust::Level;
use parking_lot::Mutex;
use uuid::{Uuid, uuid};

const ACTOR_AVA: Uuid = uuid!("11111111-1111-4111-8111-111111111111");
const PRINCIPAL_ALICE: Uuid = uuid!("22222222-2222-4222-8222-222222222222");
const INSTALL_ONE: Uuid = uuid!("33333333-3333-4333-8333-333333333333");
const INSTALL_TWO: Uuid = uuid!("44444444-4444-4444-8444-444444444444");

fn alice() -> Credential {
    Credential::new(ACTOR_AVA, PrincipalKind::User, PRINCIPAL_ALICE)
}

/// A hand-written tool and a generated CRUD collection, so both dispatch
/// branches are reachable.
fn journal_surface() -> Surface {
    let m = Manifest {
        kind: Some(Kind::App),
        name: "journal".into(),
        version: 1,
        storage: Storage {
            collections: vec![Collection {
                name: "drafts".into(),
                crud: true,
                indexes: vec![],
            }],
        },
        functions: vec![Function {
            name: "add_entry".into(),
            doc: String::new(),
        }],
        tools: vec![ToolDef {
            name: "journal.add".into(),
            function: "add_entry".into(),
            ..Default::default()
        }],
        ..Default::default()
    };
    m.validate().expect("fixture invalid");
    m.derive()
}

struct FakeInstalls(Vec<Install>);

#[async_trait]
impl Installs for FakeInstalls {
    async fn active_installs(&self, _: &Credential) -> Result<Vec<Install>, String> {
        Ok(self.0.clone())
    }
}

/// Allows exactly the (install, tool) pairs it was given, and counts how often
/// it was asked. The count is what proves both paths consult it.
struct FakeGuard {
    allow: Mutex<HashMap<String, bool>>,
    asked: Mutex<HashMap<String, usize>>,
}

fn key(install: Uuid, tool: &str) -> String {
    format!("{install}/{tool}")
}

fn guard(allowed: &[String]) -> Arc<FakeGuard> {
    Arc::new(FakeGuard {
        allow: Mutex::new(allowed.iter().map(|a| (a.clone(), true)).collect()),
        asked: Mutex::new(HashMap::new()),
    })
}

#[async_trait]
impl Guard for FakeGuard {
    async fn tool_reason(
        &self,
        _: &Credential,
        install_id: Uuid,
        tool: &str,
    ) -> Result<bool, String> {
        let k = key(install_id, tool);
        *self.asked.lock().entry(k.clone()).or_insert(0) += 1;
        Ok(self.allow.lock().get(&k).copied().unwrap_or(false))
    }
}

#[derive(Default)]
struct FakeDispatcher {
    guest_calls: Mutex<Vec<GuestCall>>,
    crud_calls: Mutex<Vec<CrudCall>>,
}

#[async_trait]
impl Dispatcher for FakeDispatcher {
    async fn call_guest(&self, call: GuestCall) -> Result<ToolResult, String> {
        self.guest_calls.lock().push(call);
        Ok(ToolResult {
            output: br#"{"via":"guest"}"#.to_vec(),
            trust: Level::Trusted,
            tainted_by: String::new(),
        })
    }
    async fn call_crud(&self, call: CrudCall) -> Result<ToolResult, String> {
        self.crud_calls.lock().push(call);
        Ok(ToolResult {
            output: br#"{"via":"crud"}"#.to_vec(),
            trust: Level::Trusted,
            tainted_by: String::new(),
        })
    }
}

fn server_with(installs: Vec<Install>, g: Arc<FakeGuard>) -> (Server, Arc<FakeDispatcher>) {
    let d = Arc::new(FakeDispatcher::default());
    let s = Server::new(Arc::new(FakeInstalls(installs)), g, d.clone());
    (s, d)
}

fn install_one() -> Install {
    Install {
        id: INSTALL_ONE,
        app: "journal".into(),
        surface: journal_surface(),
    }
}

/// THE property. Not "listing filters correctly" and not "calling denies
/// correctly", but that the two agree, over every subset of grants.
#[tokio::test]
async fn list_and_call_agree_on_every_grant_subset() {
    let surface = journal_surface();
    let names: Vec<String> = surface.tools.iter().map(|t| t.name.clone()).collect();
    assert!(
        names.len() >= 6,
        "fixture offers {} tools; too few to exercise subsets",
        names.len()
    );

    for mask in 0..(1usize << names.len()) {
        let allowed: Vec<String> = names
            .iter()
            .enumerate()
            .filter(|(i, _)| mask & (1 << i) != 0)
            .map(|(_, n)| key(INSTALL_ONE, n))
            .collect();
        let (s, _) = server_with(vec![install_one()], guard(&allowed));

        let listed = s.list_tools(&alice()).await.unwrap();
        let in_listing: Vec<String> = listed.into_iter().map(|t| t.name).collect();
        for n in &names {
            let qualified = qualified_tool_name("journal", n);
            let callable = s
                .call_tool(&alice(), &qualified, b"{}".to_vec())
                .await
                .is_ok();
            assert_eq!(
                in_listing.contains(&qualified),
                callable,
                "mask {mask}: {qualified} listed={} callable={callable}; the two disagree",
                in_listing.contains(&qualified)
            );
        }
    }
}

/// Both paths ask the predicate. If either stopped, the agreement test above
/// could still pass by both consulting the same stale cache.
#[tokio::test]
async fn both_paths_consult_the_guard() {
    let g = guard(&[key(INSTALL_ONE, "journal.add")]);
    let (s, _) = server_with(vec![install_one()], g.clone());
    s.list_tools(&alice()).await.unwrap();
    let after_list = *g
        .asked
        .lock()
        .get(&key(INSTALL_ONE, "journal.add"))
        .unwrap_or(&0);
    assert!(after_list > 0, "list_tools did not consult the guard");
    s.call_tool(&alice(), "journal.journal.add", b"{}".to_vec())
        .await
        .unwrap();
    assert!(
        *g.asked
            .lock()
            .get(&key(INSTALL_ONE, "journal.add"))
            .unwrap()
            > after_list,
        "call_tool did not consult the guard; it trusted the listing"
    );
}

/// A grant revoked between listing and calling has to bite.
#[tokio::test]
async fn revocation_between_list_and_call_bites() {
    let k = key(INSTALL_ONE, "journal.add");
    let g = guard(std::slice::from_ref(&k));
    let (s, _) = server_with(vec![install_one()], g.clone());
    let listed = s.list_tools(&alice()).await.unwrap();
    assert!(!listed.is_empty(), "nothing listed");
    g.allow.lock().insert(k, false);
    let err = s
        .call_tool(&alice(), "journal.journal.add", b"{}".to_vec())
        .await
        .err()
        .unwrap();
    assert!(
        matches!(err, McpError::UnknownTool(_)),
        "a revoked grant was served from a stale listing: {err}"
    );
}

/// Denied and non-existent are the same answer. Distinguishing them tells an
/// unauthorized caller which tools exist.
#[tokio::test]
async fn denied_and_unknown_are_indistinguishable() {
    let (s, _) = server_with(vec![install_one()], guard(&[]));
    let denied = s
        .call_tool(&alice(), "journal.journal.add", b"{}".to_vec())
        .await
        .err()
        .unwrap();
    let missing = s
        .call_tool(&alice(), "journal.does_not_exist", b"{}".to_vec())
        .await
        .err()
        .unwrap();
    assert!(matches!(denied, McpError::UnknownTool(_)), "{denied}");
    assert!(matches!(missing, McpError::UnknownTool(_)), "{missing}");
    assert_eq!(
        denied.to_string(),
        missing
            .to_string()
            .replacen("does_not_exist", "journal.add", 1)
    );
}

/// Two installs of different apps cannot see each other's tools, and the
/// qualified name is what keeps them apart.
#[tokio::test]
async fn tools_are_scoped_to_their_install() {
    let two = Install {
        id: INSTALL_TWO,
        app: "diary".into(),
        surface: journal_surface(),
    };
    // Granted on install ONE only.
    let (s, _) = server_with(
        vec![install_one(), two],
        guard(&[key(INSTALL_ONE, "journal.add")]),
    );
    let listed = s.list_tools(&alice()).await.unwrap();
    assert_eq!(listed.len(), 1);
    assert_eq!(listed[0].name, "journal.journal.add");
    // The same tool name on the other install is refused, because the grant
    // is per install rather than per name.
    let err = s
        .call_tool(&alice(), "diary.journal.add", b"{}".to_vec())
        .await
        .err()
        .unwrap();
    assert!(
        matches!(err, McpError::UnknownTool(_)),
        "a grant on one install reached another"
    );
}

/// Generated CRUD dispatches host-side with no guest, which is what lets an app
/// that is all CRUD work without a wasm module (D24).
#[tokio::test]
async fn generated_tools_dispatch_without_a_guest() {
    let (s, d) = server_with(
        vec![install_one()],
        guard(&[
            key(INSTALL_ONE, "drafts.list"),
            key(INSTALL_ONE, "journal.add"),
        ]),
    );
    s.call_tool(&alice(), "journal.drafts.list", b"{}".to_vec())
        .await
        .unwrap();
    {
        let crud = d.crud_calls.lock();
        assert_eq!(crud.len(), 1);
        assert!(
            d.guest_calls.lock().is_empty(),
            "a generated tool reached the guest dispatcher"
        );
        assert_eq!(crud[0].collection, "drafts");
        assert_eq!(crud[0].op, Op::List);
    }
    s.call_tool(&alice(), "journal.journal.add", b"{}".to_vec())
        .await
        .unwrap();
    let guest = d.guest_calls.lock();
    assert_eq!(guest.len(), 1);
    assert_eq!(guest[0].function, "add_entry");
}

/// A hidden tool is absent from both paths together, which is the same
/// guarantee rather than an exception to it.
#[tokio::test]
async fn hidden_tools_are_neither_listed_nor_callable() {
    let m = Manifest {
        kind: Some(Kind::App),
        name: "journal".into(),
        version: 1,
        storage: Storage {
            collections: vec![Collection {
                name: "drafts".into(),
                crud: true,
                indexes: vec![],
            }],
        },
        tools: vec![ToolDef {
            name: "drafts.delete".into(),
            hidden: true,
            ..Default::default()
        }],
        ..Default::default()
    };
    m.validate().expect("fixture invalid");
    let inst = Install {
        id: INSTALL_ONE,
        app: "journal".into(),
        surface: m.derive(),
    };
    // Granted, so only hiding can keep it out.
    let (s, _) = server_with(vec![inst], guard(&[key(INSTALL_ONE, "drafts.delete")]));
    for t in s.list_tools(&alice()).await.unwrap() {
        assert!(
            !t.name.ends_with("drafts.delete"),
            "a hidden tool was listed"
        );
    }
    let err = s
        .call_tool(&alice(), "journal.drafts.delete", b"{}".to_vec())
        .await
        .err()
        .unwrap();
    assert!(
        matches!(err, McpError::UnknownTool(_)),
        "a hidden tool was callable"
    );
}

/// An incomplete credential never reaches the guard, because absence of scope
/// is deny rather than a question to ask.
#[tokio::test]
async fn incomplete_credential_is_refused() {
    let g = guard(&[key(INSTALL_ONE, "journal.add")]);
    let (s, _) = server_with(vec![install_one()], g.clone());
    for cred in [
        Credential::new(Uuid::nil(), PrincipalKind::User, Uuid::nil()),
        Credential::new(ACTOR_AVA, PrincipalKind::User, Uuid::nil()),
        Credential::new(Uuid::nil(), PrincipalKind::User, PRINCIPAL_ALICE),
    ] {
        assert!(
            s.list_tools(&cred).await.is_err(),
            "list_tools accepted an incomplete credential"
        );
        assert!(
            s.call_tool(&cred, "journal.journal.add", Vec::new())
                .await
                .is_err(),
            "call_tool accepted an incomplete credential"
        );
    }
    assert!(
        g.asked.lock().is_empty(),
        "an incomplete credential reached the guard"
    );
}
